### ISSUE TITLE
Not covered `case` & `elsif`

### ISSUE THREAD

**Issue by k1r8r0wn** [deep-cover/deep-cover#70]
Hey, guys!

Have some issues in this examples (they are logically the same), I've checked all lines by debugger while running specs and 100% sure that all is covered. I tried both examples below and get this coverage results (in comments #).

What it might be?

```ruby
def method
  if first_condition?
    ...
  elsif second_condition?
    ...
  elsif third_condition? # this line is uncovered (next is covered & I've checked it by debugger)
    ...    
  end
end

def method
  case # this line is uncovered (all next ones are covered I've also checked it by debugger)
  when first_condition?
    ...
  when second_condition?
    ...    
  when third_condition?
    ...
  end
end
```

**Comment by marcandre** [comment]
Thanks for reporting.

You are right that these are equivalent. There are 4 branches, if you include the implicit `else`. Is it possible that it is that last branch that is never tested, i.e. when `first`, `second` and `third_condition?` are all falsy?

BTW, there's a switch to consider those not a problem, if you prefer

**Comment by k1r8r0wn** [comment]
@marcandre, thanks for your reply. Actually I don't need to use `else` statement in my logic, but what's wrong with `case` one?

And yep, I'm also interesting in switch, you've meant to use `ignore_uncovered` in config?

**Comment by marcandre** [comment]
Even though your `else` might do nothing, if you don't test the `else`, then if someones replaces the `elsif third_condition?` by `elsif true` by mistake (changes `third_condition?` or something else), then your tests wouldn't detect it although presumably there is a reason why your `elsif third_condition?` was there...

The switch `is_case_implicit_else` should make your second example pass. There is no switch for the first example, but it would be easy to add.

**Comment by k1r8r0wn** [comment]
I see, but I've already tried to add `else` with `nil` inside just for test & guess what - got the uncovered line in this `else` statement.

Also tried `is_case_implicit_else`, but it won't work with uncovered `case`.

Was it correctly defined?

```ruby
DeepCover.configure do
  ignore_uncovered :is_case_implicit_else
end
```

**Comment by marcandre** [comment]
Sorry, my bad, should be: `ignore_uncovered :case_implicit_else`

You need to understand that the idea between branch coverage... If I have

```ruby
def my_method
  # ...
  do_something if condition?
  # ...
end
```

To be actually covered, `my_method` needs to be run at least once when `condition?` is false and at least once when `condition?` is true. If you only test with `condition?` being true, then either there's an error in your code (the `if condition?` is not needed) or there's a hole in your coverage. HTH

**Comment by k1r8r0wn** [comment]
Yeah, I understand, generally speaking I need to use else statement & test it too, even if I don't need it at all) But what's wrong with `case` then?

**Comment by marcandre** [comment]
Did you try with `ignore_uncovered :case_implicit_else` (without the `is_`)?

**Comment by k1r8r0wn** [comment]
Yep, recently

Interesting story that it actually works in other parts of app, where it is this way:

```ruby
def working_case
 case smt
 when 'a'
  ...
 when 'b'
  ...
  end
end
```

**Comment by marcandre** [comment]
If you look at the HTML report, can you tell me what the message is if you hover?
Otherwise I'd need an actual code example to check it out...

**Comment by k1r8r0wn** [comment]
One moment, @marcandre. The message is - `implicit else not covered`. The problem is that I don't use `else` here.

**Comment by k1r8r0wn** [comment]
Is it possible to disable this check for particular methods in code, not in whole app?

**Comment by marcandre** [comment]
If you run `deep-cover exec --ignore-case-implicit-else` you also get this?

**Comment by k1r8r0wn** [comment]
Yeah, this flag doesn't help

**Comment by marcandre** [comment]
Are you running the latest version of `deep-cover`?

I made a simple example for myself:
```ruby
# ~/temp/case/case.rb:
Foo = Struct.new(:first, :second) do
  def method
    case
    when first
      puts "here"
      self.first = nil
    when second
      puts "there"
      self.second = nil
    end
  end
end


foo = Foo.new(true, true)
foo.method
foo.method
```

```sh
deep-cover clone --ignore-case-implicit-else -C ~/temp/case/ ruby case.rb
Cloning... /Users/mal/temp/c/case
Rewriting: |=====================================================================================================| Time: 00:00:00
here
there
|  Path   |    Nodes     |   %    |   Chars   |   %    |  Branches  |   %    |
+---------+--------------+--------+-----------+--------+------------+--------+
| case.rb | 36 [+1] / 37 | 100.00 | 148 / 148 | 100.00 | 0 [+1] / 1 | 100.00 |

Overall: 100.0%
```

The `[+1]` above is the ignored implicit else node. With `--no-ignore-case-implicit-else` instead there are no `[+x]`, overall is 97%, etc.

**Comment by k1r8r0wn** [comment]
Hmm, I see, @marcandre, all is pretty fine. Maybe the issue is that I use `deep-cover` with `simplecov`, but I'm pretty sure that I've tuned it the right way, with the help of info provided in readme...

**Comment by marcandre** [comment]
Oh, right. There might actually be an issue with our line coverage, let me investigate further

**Comment by k1r8r0wn** [comment]
Much appreciated @marcandre!

**Comment by marcandre** [comment]
Hey, I'm sorry I didn't click more rapidly on this one. You were right, anything that had to do with "per-line" analysis was disregarding the options.

As the API should be stable I released the v1.0.0 🎉 

Thanks again for the report

**Comment by k1r8r0wn** [comment]
Glad to help, thank you too!
